packages/nx/src/tasks-runner/run-command.ts

```diff
@@ -198,6 +209,171 @@ export async function runCommand(
+}> {
+  let taskGraph = createTaskGraphAndRunValidations(
+    projectGraph,
+    extraTargetDependencies ?? {},
+    projectNames,
+    nxArgs,
+    overrides,
+    extraOptions
+  );
+
+  if (process.env.NX_ENABLE_SYNC_GENERATORS !== 'true') {
+    return { projectGraph, taskGraph };
+  }
+
+  // collect unique syncGenerators from the tasks
+  const uniqueSyncGenerators = new Set<string>();
+  for (const { target } of Object.values(taskGraph.tasks)) {
+    const { syncGenerators } =
+      projectGraph.nodes[target.project].data.targets[target.target];
+    if (!syncGenerators) {
+      continue;
+    }
+
+    for (const generator of syncGenerators) {
+      uniqueSyncGenerators.add(generator);
+    }
+  }
+
+  if (!uniqueSyncGenerators.size) {
+    // There are no sync generators registered in the tasks to run
+    return { projectGraph, taskGraph };
+  }
+
+  const syncGenerators = Array.from(uniqueSyncGenerators);
+  const results = await getSyncGeneratorChanges(syncGenerators);
+  if (!results.length) {
+    // There are no changes to sync, workspace is up to date
+    return { projectGraph, taskGraph };
+  }
+
+  const outOfSyncTitle = 'The workspace is out of sync';
+  const resultBodyLines = syncGeneratorResultsToMessageLines(results);
+  const fixMessage =
+    'You can manually run `nx sync` to update your workspace or you can set `sync.applyChanges` to `true` in your `nx.json` to apply the changes automatically when running tasks.';
+  const willErrorOnCiMessage = 'Please note that this will be an error on CI.';
+
+  if (isCI() || !process.stdout.isTTY) {
+    // If the user is running in CI or is running in a non-TTY environment we
+    // throw an error to stop the execution of the tasks.
+    throw new Error(
+      `${outOfSyncTitle}\n${resultBodyLines.join('\n')}\n${fixMessage}`
+    );
+  }
+
+  if (nxJson.sync?.applyChanges === false) {
+    // If the user has set `sync.applyChanges` to `false` in their `nx.json`
+    // we don't prompt the them and just log a warning informing them that
+    // the workspace is out of sync and they have it set to not apply changes
+    // automatically.
+    output.warn({
+      title: outOfSyncTitle,
+      bodyLines: [
+        ...resultBodyLines,
+        'Your workspace is set to not apply changes automatically (`sync.applyChanges` is set to `false` in your `nx.json`).',
+        willErrorOnCiMessage,
+        fixMessage,
+      ],
+    });
+    return { projectGraph, taskGraph };
+  }
+
+  output.warn({
+    title: outOfSyncTitle,
+    bodyLines: [
+      ...resultBodyLines,
+      nxJson.sync?.applyChanges === true
+        ? 'Proceeding to sync the changes automatically (`sync.applyChanges` is set to `true` in your `nx.json`).'
+        : willErrorOnCiMessage,
+    ],
+  });
+
+  const applyChanges =
+    nxJson.sync?.applyChanges === true ||
+    (await promptForApplyingSyncGeneratorChanges());
+
+  if (applyChanges) {
+    const spinner = ora('Syncing workspace configuration...');
```

Comment:
Can we keep it more general as the "workspace" is out of sync rather than the "workspace configuration" is out of sync?